### ISSUE TITLE
Asgard's Wrath - Index Controller "grab" issues.

### ISSUE THREAD

**Issue by Scope666** [LibreVR/Revive#1355]
Index controller is having "grab" issues in Asgard's Wrath.  Sometimes while swinging the sword it will just fall out of your hands despite the pressure used to hold it.  Also too easy to accidentally grab when it's holstered.

**To Reproduce**
Grab sword, start swinging, try to keep in hand.  Move left hand near holstered weapon.

**Expected behavior**
Index should know whether you intend to hold the item or not based on pressure and amount of fingers wrapped around controller.

**Environment (please complete the following information):**
 - App: Asgard's Wrath
 - OS: Win 10  build 1903
 - Headset: Index w/ Index Controllers

**Versions (please complete the following information):**
 - Revive: 1.8.2.4    (similar with 1.8.1)
 - SteamVR: 1.7.15

**Additional context**
Tried experimenting with different options in Steam VR Bindings interface but couldn't find anything to improve the situation.  The current implementation for Revive with Index Controllers doesn't allow any sensitivity adjustments within this interface.

**Comment by ahmanwoods** [comment]
Are you dropping your sword during combat? Certain enemy moves can disarm you AFAIK.

**Comment by Armada651** [comment]
There's the option of either binding the HandTrigger to Force or binding it
to a binary Grab which has input sensitivity.

To get the Force/Grab input you first need to delete the Grip input and
then add a new one that you bind to the Grip axis.

**Comment by hookmanuk** [comment]
Yes, I created a controller binding a few days ago for this under my hookman steam user.

It makes the grab more "sticky" so you can hold it quite loosely once you have grabbed the sword. It also makes you need more force to grab it in the first place, which stops you accidentally grabbing a holstered weapon.

One thing I have noticed is everytime I run it tells me my binding was created for an older version of the game in steam VR, so I have to manually reselect it. I'm not sure why as neither the game (as far as I know) has been patched, nor had I upgraded revive. Not sure if this is a known issue?

**Comment by Scope666** [comment]
> Are you dropping your sword during combat? Certain enemy moves can disarm you AFAIK.

Definitely wasn't combat only (I know what you're talking about, when you try to parry certain attacks they can knock the sword out of your hand)

It almost seems intermittent, like other controller movements tend to make it happen.  When I was experimenting with Binary Grab it affected the ability to throw the axe.

**Comment by Armada651** [comment]
If you want to use the binary grab mode you can simply do so in the SteamVR input settings. The default binding is an analog grip value as that more accurately represents the hand trigger on an Oculus Touch

**Comment by Scope666** [comment]
Perhaps the default settings for SteamVR input needs to be tweaked for the Index.  The Touch controller has that simple grab button, it would be nice to have a "default" setting that just works for most if not all Oculus titles.

This is what I came up with for Index that improved things for me as an example:

![Screenshot_1](https://user-images.githubusercontent.com/31927547/67100856-74e00800-f18e-11e9-87c2-3ff56e6ac1eb.jpg)

**Comment by helifax** [comment]
> Perhaps the default settings for SteamVR input needs to be tweaked for the Index. The Touch controller has that simple grab button, it would be nice to have a "default" setting that just works for most if not all Oculus titles.
> 
> This is what I came up with for Index that improved things for me as an example:
> 
> ![Screenshot_1](https://user-images.githubusercontent.com/31927547/67100856-74e00800-f18e-11e9-87c2-3ff56e6ac1eb.jpg)

This is just a binary operation: Hand open or hand close! No "in-between" states. The current implementation shows the "analogue"  value read from the sensors. If you pay attention you can see "how your hand turns into a fist", rather than going from fully open to "fist mode".

There is no issue in Revive. I think the issue is with us not being able to calibrate the sensitivity of the Index Controller Sensors (Globally!) I also find them way to sensitive, but re-binding keys, will not solve the fundamental issue that are seeing!

**Comment by helifax** [comment]
One thing that I noticed:
Followers don't respond back to any gestures. For example, when using the Oculus Rift if I do a "thumbs-up" the follower will respond in kind.
I tried this on Revive with my Vive Pro and I couldn't get the follower to respond. I believe there is something in the Input System, that Revive is missing, but I can't figure what exactly.

Any thoughts? :)

**Comment by Scope666** [comment]
After reading your post I went back and tried the "Official" bindings for Revive.  You're right that you can open / close your hand variably, but the game is basically unplayable this way, as it's too easy to accidentally grab either weapon on your hip.

I know Index has TWO ways it tells if you're holding something.  The POSITION of the 3 fingers, and the FORCE that they are squeezing the grip with.  As far as I can tell the default binding seems to ignore FORCE and only care about POSITION.  The problem with position is even one finger close to the grip is enough to trigger a grab operation in Asgard's Wrath.

That's why I'm suggesting that the default binding for Revive might need to be investigated...

**Comment by Scope666** [comment]
> One thing that I noticed:
> Followers don't respond back to any gestures. For example, when using the Oculus Rift if I do a "thumbs-up" the follower will respond in kind.
> I tried this on Revive with my Vive Pro and I couldn't get the follower to respond. I believe there is something in the Input System, that Revive is missing, but I can't figure what exactly.
> 
> Any thoughts? :)

I tested and I can get the Follower to respond with both my binding and the Official one on the Index, it might be something specific to the Vive Pro (Wand) controller binding?

**Comment by helifax** [comment]
> > One thing that I noticed:
> > Followers don't respond back to any gestures. For example, when using the Oculus Rift if I do a "thumbs-up" the follower will respond in kind.
> > I tried this on Revive with my Vive Pro and I couldn't get the follower to respond. I believe there is something in the Input System, that Revive is missing, but I can't figure what exactly.
> > Any thoughts? :)
> 
> I tested and I can get the Follower to respond with both my binding and the Official one on the Index, it might be something specific to the Vive Pro (Wand) controller binding?

I am using the Index Controllers with the Vive Pro. I even didn't try the wands yet (as they mostly not suited for games, unfortunately).
I can "High-Five" and "Fist-bump" when the followers want, but randomly, if I just "thumbs-up" they ignore me... (Maybe I am doing it wrong, which is probably what I am doing...)

**Comment by helifax** [comment]
> After reading your post I went back and tried the "Official" bindings for Revive. You're right that you can open / close your hand variably, but the game is basically unplayable this way, as it's too easy to accidentally grab either weapon on your hip.
> 
> I know Index has TWO ways it tells if you're holding something. The POSITION of the 3 fingers, and the FORCE that they are squeezing the grip with. As far as I can tell the default binding seems to ignore FORCE and only care about POSITION. The problem with position is even one finger close to the grip is enough to trigger a grab operation in Asgard's Wrath.
> 
> That's why I'm suggesting that the default binding for Revive might need to be investigated...

You are correct! If only we could just calibrate the sensitivity of the sensors... the whole problem would be solved 👍 
I tried to bind both the analogue and the binary to the "grip" part of the controller, but the SteamVR UI doesn't allow for this :(

**Comment by BOLL7708** [comment]
The only thing I tried to do to fix this was make it so every finger need to touch the handle to grab, but could not find how to do it in the bindings. 

Whatever change I did, I noticed I ruined throwing, maybe by using the force sensor, if I reverted to Revive defaults throwing would work well again. So right now, I do run around super easily grabbing anything with just one finger on the handle, but, throwing is at least possible 😬

**Comment by Scope666** [comment]
Try my settings from the screen shot above (tweak to your liking after testing)  I can throw just fine with them.

**Comment by Scope666** [comment]
I see this just dropped today, nice to see the Valve guys paying attention to Revive!

https://steamcommunity.com/games/250820/announcements/detail/1657814577467632328
